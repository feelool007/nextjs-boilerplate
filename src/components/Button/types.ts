import { Omit } from "@material-ui/core";
import { ButtonProps } from "@material-ui/core/Button";

import { PThemeProvider } from "../ThemeProvider";

export interface PButton extends PThemeProvider, Omit<ButtonProps, "color"> {}
