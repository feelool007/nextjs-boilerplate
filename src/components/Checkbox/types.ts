import { Omit } from "@material-ui/core";
import { CheckboxProps } from "@material-ui/core/Checkbox";
import { FormControlLabelProps } from "@material-ui/core/FormControlLabel";

import { PThemeProvider } from "../ThemeProvider";

export interface PCheckbox extends PThemeProvider, Omit<CheckboxProps, "color"> {
  value?: string;
  label: string;
  formControlLabelProps?: FormControlLabelProps
}
