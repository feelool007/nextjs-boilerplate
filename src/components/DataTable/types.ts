import { WithStyles, Omit } from "@material-ui/core";
import { TableProps } from "@material-ui/core/Table";
import { LinkProps } from "react-csv/components/Link";

import { toolbarStyles, dataTableStyles } from "./styles";
import { PInput } from "../Input";

export type TableDataHeader<T = { [key: string]: any }> = {
  column: keyof T;
  label: string;
};

export type SortDirectionKeys = "asc" | "desc";

export interface PDataTable
  extends WithStyles<typeof dataTableStyles>,
    Omit<TableProps, "classes" | "title" | "onSelect"> {
  headers: Array<TableDataHeader>;
  data: Array<Object>;
  sort?: boolean;
  hover?: boolean;
  defaultSortDirection?: SortDirectionKeys;
  defaultSortBy?: string;
  select?: boolean;
  selectBy?: string;
  pick?: boolean;
  pickBy?: string;
  toolbar?: boolean;
  title?: React.ReactNode;
  search?: boolean;
  csv?: boolean;
  rowsPerPageOptions?: Array<number>;
  actions?: Array<React.ReactElement>;
  actionHeaders?: Array<string>;
  onSelect?: (selected: Array<any>) => void;
  onPick?: (picked: any) => void;
}

export interface SDataTable {
  sortBy: any;
  sortDirection: SortDirectionKeys;
  page: number;
  rowsPerPage: number;
  selected: Array<any>;
  picked: any;
  searchValue: string;
}

export interface PTableToolbar extends WithStyles<typeof toolbarStyles> {
  title?: React.ReactNode;
  search?: boolean;
  SearchProps: Omit<PSearch, "classes">;
  csv?: boolean;
  CSVDownloadProps: Omit<PCSVDownload, "classes">;
}

export interface STableToolbar {}

export interface PSearch extends WithStyles<typeof toolbarStyles>, Omit<PInput, "classes"> {
  onClear: () => void;
}

export interface SSearch {
  open: boolean;
  disableUnderline: boolean;
}

export interface PCSVDownload extends WithStyles<typeof toolbarStyles>, LinkProps {}
