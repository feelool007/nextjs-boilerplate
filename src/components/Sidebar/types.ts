import { WithStyles } from "@material-ui/core";
import { SvgIconProps } from "@material-ui/core/SvgIcon";

import { sidebarStyles } from "./styles";

export interface Page {
  name: string,
  to: string
}

export interface PageGroup {
  icon: React.ComponentType<SvgIconProps>,
  groupName?: string,
  pages: Page[]
}

export interface PSidebarMenu extends WithStyles<typeof sidebarStyles> {
  mini?: boolean,
  pageGroup: PageGroup
}

export interface SSidebarMenu {
  openCollapse: boolean,
  openPopover: boolean
}

export interface PSidebar extends WithStyles<typeof sidebarStyles> {
  brand?: string,
  mini?: boolean,
  rwdOpen?: boolean,
  pageGroups: PageGroup[],
  onToggleRwd: () => void
}
