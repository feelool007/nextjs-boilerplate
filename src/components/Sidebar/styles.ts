import { createStyles, Theme } from "@material-ui/core";

export const sidebarStyles = (theme: Theme) => createStyles({
  rootBasic: {
    backgroundColor: "#2c2c2c",
    position: "relative",
    height: "100vh",
    overflowX: "hidden",
    transition: "width 0.2s"
  },
  rootMini: {
    width: 60
  },
  rootNormal: {
    width: 260
  },
  brandContainer: {
    width: "100%",
    overflow: "hidden"
  },
  brand: {
    fontSize: 20,
    fontWeight: "bold",
    color: "rgb(235, 235, 235)",
    textAlign: "center",
    paddingBottom: theme.spacing.unit * 1,
    borderBottom: "0.3mm solid white"
  },
  brandMini: {
    fontSize: 20,
    textAlign: "center",
    color: "rgb(235, 235, 235)",
  },
  menuRoot: {
    padding: 0
  },
  menu: {
    fontSize: 16,
    height: 55
  },
  popperMenu: {
    backgroundColor: "#424242",
    width: 240,
    paddingTop: theme.spacing.unit * 1,
    paddingBottom: theme.spacing.unit * 1,
  },
  menuOpen: {
    color: "#00ACC1"
  },  
  menuClose: {
    color: "rgb(185, 185, 185)",
    "&:hover": {
      color: "rgb(255, 255, 255)"
    }
  },
  menuText: {
    marginLeft: theme.spacing.unit * 2
  },
  menuItem: {
    height: 40
  },
  menuItemText: {
    fontSize: 14,
    color: "rgb(185, 185, 185)",
    paddingLeft: "1.5em",
    "&:hover": {
      color: "rgb(255, 255, 255)"
    }
  },
  menuItemHighlight: {
    backgroundColor: "#00bcd4",
    width: "85%",
    marginLeft: theme.spacing.unit * 2,
    border: "1px solid #00838f",
    borderRadius: 8,
    "&:hover": {
      backgroundColor: "#00bcd4"
    }
  },
  menuItemTextHighlight: {
    color: "white",
    fontWeight: "bold"
  },
  icon: {
    fontSize: 16
  }
})
