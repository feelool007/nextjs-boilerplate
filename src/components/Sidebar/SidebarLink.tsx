import React from "react";
import { withStyles, ListItem, Grid } from "@material-ui/core";
import { People } from "@material-ui/icons";
import Link from "next/link";
import { withRouter } from "next/router";
import classNames from "classnames";

import { PSidebarLink } from "./types";
import { sidebarStyles } from "./styles";

class SidebarLink extends React.Component<PSidebarLink> {
  render = () => {
    const { page, router, mini, classes } = this.props;
    const { icon: Icon } = page;
    return (
      <Grid container justify="center">
        <Link href={page.to}>
          <ListItem
            button
            className={classNames({
              [classes.menuItem]: true,
              [classes.menuItemHighlight]: page.to === router.pathname
            })}
          >
            {Icon ? (
              <Icon className={classes.icon} />
            ) : (
              // placeholder, only use in mini mode
              !mini && <People className={classNames(classes.icon, classes.visibleHidden)} />
            )}
            <div className={classes.flex}>
              <p
                className={classNames({
                  [classes.menuItemText]: true,
                  [classes.menuItemTextHighlight]: page.to === router.pathname
                })}
              >
                {page.name}
              </p>
            </div>
          </ListItem>
        </Link>
      </Grid>
    );
  };
}

// @ts-ignore
export default withStyles(sidebarStyles)(withRouter(SidebarLink));
