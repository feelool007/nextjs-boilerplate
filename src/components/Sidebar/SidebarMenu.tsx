import React from "react";
import { List, ListItem, ListItemText, Collapse, Paper, Popper, withStyles } from "@material-ui/core";
import { KeyboardArrowDown, KeyboardArrowUp } from "@material-ui/icons";
import Link from "next/link";
import { withRouter } from "next/router";
import classNames from "classnames";

import { PSidebarMenu, SSidebarMenu } from "./types";
import { sidebarStyles } from "./styles";

class SidebarMenu extends React.Component<PSidebarMenu, SSidebarMenu> {
  static defaultProps = {
    mini: false
  };

  private __ref: React.RefObject<HTMLButtonElement>;

  constructor(props: PSidebarMenu) {
    super(props);
    this.state = {
      openCollapse: false,
      openPopover: false
    };
    this.__ref = React.createRef();
  }

  handleToggleCollapse = () => {
    this.setState(prevState => ({
      openCollapse: !prevState.openCollapse
    }));
  };

  handleOpenPopover = () => {
    this.setState({
      openPopover: true
    });
  };

  handleClosePopover = () => {
    this.setState({
      openPopover: false
    });
  };

  handleBlur = () => {
    const listItems = document.getElementsByName("li-item");
    listItems.forEach(el => el.blur());
  };

  handleGenerateMenu = () => {
    const { router, classes } = this.props;
    const { pages } = this.props.pageGroup;
    return pages.map((d, index) => {
      return (
        <Link key={index} href={d.to}>
          <ListItem
            button
            name="li-item"
            className={classNames({
              [classes.menuItem]: true,
              [classes.menuItemHighlight]: d.to === router.pathname
            })}
            onClick={this.handleBlur}
          >
            <ListItemText
              primary={d.name}
              classes={{
                primary: classNames({
                  [classes.menuItemText]: true,
                  [classes.menuItemTextHighlight]: d.to === router.pathname
                })
              }}
            />
          </ListItem>
        </Link>
      );
    });
  };

  render = () => {
    const { pageGroup, mini, classes } = this.props;
    const { icon: Icon, groupName } = pageGroup;
    const { openCollapse, openPopover } = this.state;
    return mini ? (
      <List onMouseEnter={this.handleOpenPopover} onMouseLeave={this.handleClosePopover} className={classes.menuRoot}>
        <ListItem
          button
          divider
          buttonRef={this.__ref}
          className={classNames({
            [classes.menu]: true,
            [classes.menuOpen]: openCollapse,
            [classes.menuClose]: !openCollapse
          })}
        >
          <Icon className={classes.icon} />
        </ListItem>
        <Popper open={openPopover} anchorEl={this.__ref.current} placement="right-start">
          <Paper className={classes.popperMenu}>{this.handleGenerateMenu()}</Paper>
        </Popper>
      </List>
    ) : (
      <List className={classes.menuRoot}>
        <ListItem
          button
          divider
          onClick={this.handleToggleCollapse}
          className={classNames({
            [classes.menu]: true,
            [classes.menuOpen]: openCollapse,
            [classes.menuClose]: !openCollapse
          })}
        >
          <Icon className={classes.icon} />
          <div className={classes.flex}>
            <p className={classes.menuText}>{groupName}</p>
          </div>
          {openCollapse ? <KeyboardArrowUp className={classes.icon} /> : <KeyboardArrowDown className={classes.icon} />}
        </ListItem>
        <Collapse in={openCollapse}>{this.handleGenerateMenu()}</Collapse>
      </List>
    );
  };
}

// There is a bug in withRouter type checking. Keep it quiet until fixed.
// @ts-ignore
export default withStyles(sidebarStyles)(withRouter(SidebarMenu));
