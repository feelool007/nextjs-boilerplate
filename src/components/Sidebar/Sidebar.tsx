import React from "react";
import Link from "next/link";
import { Drawer, withStyles, List, ListItem, ListItemText, Hidden } from "@material-ui/core";
import { Home } from "@material-ui/icons";
import classNames from "classnames";
import PerfectScrollbar from "react-perfect-scrollbar";
import "react-perfect-scrollbar/dist/css/styles.css"

import { PSidebar } from "./types";
import { sidebarStyles } from "./styles";
import SidebarMenu from "./SidebarMenu";

class Sidebar extends React.Component<PSidebar> {
  static defaultProps = {
    brand: "",
    mini: false,
    rwdOpen: false
  };

  handleGenerateContent = () => {
    const { mini, brand, pageGroups, classes } = this.props;
    return (
      <PerfectScrollbar>
        <List className={classes.brandContainer}>
          <Link href="/">
            <ListItem button>
              {mini ? (
                <Home className={classes.brandMini} />
              ) : (
                <ListItemText
                  primary={brand}
                  classes={{
                    primary: classes.brand
                  }}
                />
              )}
            </ListItem>
          </Link>
        </List>
        {pageGroups.map((d, index) => {
          return <SidebarMenu mini={mini} key={index} pageGroup={d} />;
        })}
      </PerfectScrollbar>
    );
  };

  render = () => {
    const { mini, rwdOpen, onToggleRwd, classes } = this.props;
    return (
      <React.Fragment>
        <Hidden mdDown>
          <Drawer
            open
            variant="permanent"
            classes={{
              paper: classNames({
                [classes.rootBasic]: true,
                [classes.rootMini]: mini,
                [classes.rootNormal]: !mini
              })
            }}
          >
            {this.handleGenerateContent()}
          </Drawer>
        </Hidden>
        <Hidden lgUp>
          <Drawer
            open={rwdOpen}
            variant="temporary"
            classes={{
              paper: classNames({
                [classes.rootBasic]: true,
                [classes.rootNormal]: true
              })
            }}
            transitionDuration={250}
            onClose={onToggleRwd}
          >
            {this.handleGenerateContent()}
          </Drawer>
        </Hidden>
      </React.Fragment>
    );
  };
}

export default withStyles(sidebarStyles)(Sidebar);
