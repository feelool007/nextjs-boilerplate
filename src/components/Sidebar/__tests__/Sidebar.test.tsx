import { createShallow } from "@material-ui/core/test-utils";
import { ShallowWrapper } from "enzyme";
import { Omit } from "@material-ui/core";
import { Edit } from "@material-ui/icons";

import { PageGroup, PSidebar } from "../types";
import Sidebar from "../Sidebar";
import SidebarMenu from "../SidebarMenu";

let wrapper: ShallowWrapper;
let shallow = createShallow({ dive: true });

const pages: Array<PageGroup> = [
  { icon: Edit, groupName: "Index", pages: [{ name: "home", to: "/" }] },
  { icon: Edit, groupName: "About", pages: [{ name: "about", to: "about" }] }
];

describe("Sidebar UI Test", () => {
  let props: Omit<PSidebar, "classes">;

  beforeEach(() => {
    props = {
      pageGroups: pages,
      onToggleRwd: jest.fn()
    };
    wrapper = shallow(<Sidebar {...props} />);
  });

  it("component snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should contain 4 SidebarMenu", () => {
    expect(wrapper.find(SidebarMenu)).toHaveLength(4);
  });
});
