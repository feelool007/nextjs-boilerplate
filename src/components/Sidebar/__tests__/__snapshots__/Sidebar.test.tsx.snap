// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar UI Test component snapshot 1`] = `
<Fragment>
  <Hidden
    implementation="js"
    lgDown={false}
    lgUp={false}
    mdDown={true}
    mdUp={false}
    smDown={false}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <WithStyles(ForwardRef(Drawer))
      classes={
        Object {
          "paper": "Sidebar-rootBasic-1 Sidebar-rootNormal-3",
        }
      }
      open={true}
      variant="permanent"
    >
      <ScrollBar
        className=""
        component="div"
        containerRef={[Function]}
      >
        <WithStyles(ForwardRef(List))
          className="Sidebar-brandContainer-4"
        >
          <Link
            href="/"
          >
            <WithStyles(ForwardRef(ListItem))
              button={true}
            >
              <WithStyles(ForwardRef(ListItemText))
                classes={
                  Object {
                    "primary": "Sidebar-brand-5",
                  }
                }
                primary=""
              />
            </WithStyles(ForwardRef(ListItem))>
          </Link>
        </WithStyles(ForwardRef(List))>
        <WithStyles(withRouter(SidebarMenu))
          key="0"
          mini={false}
          pageGroup={
            Object {
              "groupName": "Index",
              "icon": Object {
                "$$typeof": Symbol(react.memo),
                "compare": null,
                "displayName": "EditIcon",
                "muiName": "SvgIcon",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
              },
              "pages": Array [
                Object {
                  "name": "home",
                  "to": "/",
                },
              ],
            }
          }
        />
        <WithStyles(withRouter(SidebarMenu))
          key="1"
          mini={false}
          pageGroup={
            Object {
              "groupName": "About",
              "icon": Object {
                "$$typeof": Symbol(react.memo),
                "compare": null,
                "displayName": "EditIcon",
                "muiName": "SvgIcon",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
              },
              "pages": Array [
                Object {
                  "name": "about",
                  "to": "about",
                },
              ],
            }
          }
        />
      </ScrollBar>
    </WithStyles(ForwardRef(Drawer))>
  </Hidden>
  <Hidden
    implementation="js"
    lgDown={false}
    lgUp={true}
    mdDown={false}
    mdUp={false}
    smDown={false}
    smUp={false}
    xlDown={false}
    xlUp={false}
    xsDown={false}
    xsUp={false}
  >
    <WithStyles(ForwardRef(Drawer))
      classes={
        Object {
          "paper": "Sidebar-rootBasic-1 Sidebar-rootNormal-3",
        }
      }
      onClose={[Function]}
      open={false}
      transitionDuration={250}
      variant="temporary"
    >
      <ScrollBar
        className=""
        component="div"
        containerRef={[Function]}
      >
        <WithStyles(ForwardRef(List))
          className="Sidebar-brandContainer-4"
        >
          <Link
            href="/"
          >
            <WithStyles(ForwardRef(ListItem))
              button={true}
            >
              <WithStyles(ForwardRef(ListItemText))
                classes={
                  Object {
                    "primary": "Sidebar-brand-5",
                  }
                }
                primary=""
              />
            </WithStyles(ForwardRef(ListItem))>
          </Link>
        </WithStyles(ForwardRef(List))>
        <WithStyles(withRouter(SidebarMenu))
          key="0"
          mini={false}
          pageGroup={
            Object {
              "groupName": "Index",
              "icon": Object {
                "$$typeof": Symbol(react.memo),
                "compare": null,
                "displayName": "EditIcon",
                "muiName": "SvgIcon",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
              },
              "pages": Array [
                Object {
                  "name": "home",
                  "to": "/",
                },
              ],
            }
          }
        />
        <WithStyles(withRouter(SidebarMenu))
          key="1"
          mini={false}
          pageGroup={
            Object {
              "groupName": "About",
              "icon": Object {
                "$$typeof": Symbol(react.memo),
                "compare": null,
                "displayName": "EditIcon",
                "muiName": "SvgIcon",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
              },
              "pages": Array [
                Object {
                  "name": "about",
                  "to": "about",
                },
              ],
            }
          }
        />
      </ScrollBar>
    </WithStyles(ForwardRef(Drawer))>
  </Hidden>
</Fragment>
`;
